juju/model.py

```diff
@@ -2108,23 +2108,18 @@ def resolve(self, reference):
             will be 0 and separate AddUnitChanges will be used.  For Kubernetes
             models, this will be used to scale the application.
             (Only given on Juju 2.5+)
-
-        :param placement string:
-            Placement holds the placement directive for units of this
-            application.  Only applicable for Kubernetes models.
-            (Only given on Juju 2.5+)
         """
         # resolve indirect references
         charm = self.resolve(charm)
 
         if len(args) == 1:
             # Juju 2.4 and below only sends the resources
             resources = args[0]
-            devices, num_units, placement = None, None, None
+            devices, num_units = None, None
         else:
             # Juju 2.5+ sends devices before resources, as well as num_units
-            # and placement
-            devices, resources, num_units, placement = args
+            # There might be placement but we need to ignore that.
+            devices, resources, num_units = args[:3]
```

Comment:
Model.deploy is using:
    async def deploy(
            self, entity_url, application_name=None, bind=None, budget=None,
            channel=None, config=None, constraints=None, force=False,
            num_units=1, plan=None, resources=None, series=None, storage=None,
            to=None, devices=None):

BundleHandler.deploy isn't. Is it plausible to change bundlehandler to use the named argument form instead of strictly positional?

I don't see where in the code it is calling \.deploy, so I'm guessing it is something at a different level?

I see places in examples, etc, but nothing under juju/* that is calling deploy() directly.

*args is always going to be a problem as you have to pass exactly only 1 thing in the 5th place, and we should never ever change what that thing is. And if you are removing placement in this patch, and we ever need to *add* something in the future, its going to break because then both 'placement' and 'newthing' end up in position arg[3].

Reply:
`BundleHandler.deploy()` is dispatched dynamically based on the `BundleChangeResult` objects returned by `BundleFacade.GetChanges()`:

https://github.com/juju/python-libjuju/blob/2acbdc400a98e97882b902d922383bd64e4fb101/juju/model.py#L1963-L1967

It might make it more clear to explicitly enumerate the mapping between the `step.method` value and the actual method being called, but Python doesn't have a switch / select statement, so it would have to be done with a large `if` / `elif` block or a `dict` lookup table.

Unfortunately, the positional arg usage comes from the `BundleChangeResult` structure and I think is ultimately dictated by the GUI.  And since JavaScript doesn't support named arguments, I don't know that we can realistically change it.  But if that is an option, I'm definitely for it.